frontend/connection/src/ForwardMessageCache.test.ts

```diff
@@ -176,10 +170,82 @@ test("removes expired messages", () => {
   expect(getCachedMessage(msg.hash)).toEqual(msg)
 
   // Increment our age. Our message should still exist.
-  cache.incrementRunCount(1)
+  cache.incrementRunCount(1, [])
   expect(getCachedMessage(msg.hash)).toEqual(msg)
 
   // Bump our age over the expiration threshold.
-  cache.incrementRunCount(1)
+  cache.incrementRunCount(1, [])
   expect(getCachedMessage(msg.hash)).toBeUndefined()
 })
+
+test("only expires messages with matching fragment IDs", () => {
+  const { cache, getCachedMessage } = createCache()
+
+  // Create messages with different fragment IDs
+  const msg1 = createForwardMsgWithFragment("msg1", "fragment-1")
+  const msg2 = createForwardMsgWithFragment("msg2", "fragment-2")
+  const msg3 = createForwardMsg("msg3") // Message without fragment ID
+
+  const encodedMsg1 = ForwardMsg.encode(msg1).finish()
+  const encodedMsg2 = ForwardMsg.encode(msg2).finish()
+  const encodedMsg3 = ForwardMsg.encode(msg3).finish()
+
+  // Add messages to the cache
+  // @ts-expect-error accessing into internals for testing
+  cache.maybeCacheMessage(msg1, encodedMsg1)
+  // @ts-expect-error accessing into internals for testing
+  cache.maybeCacheMessage(msg2, encodedMsg2)
+  // @ts-expect-error accessing into internals for testing
+  cache.maybeCacheMessage(msg3, encodedMsg3)
+
+  // Verify all messages are in cache
+  expect(getCachedMessage("msg1")).toEqual(msg1)
+  expect(getCachedMessage("msg2")).toEqual(msg2)
+  expect(getCachedMessage("msg3")).toEqual(msg3)
+
+  // Increment run count with fragment-1 - should only consider msg1 for expiration
+  cache.incrementRunCount(1, ["fragment-1"])
+
+  // Non-matching fragment IDs and messages without fragment ID should still be cached
+  expect(getCachedMessage("msg1")).toEqual(msg1) // Not expired yet (age = 1)
+  expect(getCachedMessage("msg2")).toEqual(msg2) // Not considered for expiration
+  expect(getCachedMessage("msg3")).toEqual(msg3) // Not considered for expiration
+
+  // Another increment with fragment-1 - should expire fragment-1 message
+  cache.incrementRunCount(1, ["fragment-1"])
+
+  // fragment1 (with fragment-1) should now be expired, others should remain
+  expect(getCachedMessage("msg1")).toBeUndefined() // Now expired (age > 1)
+  expect(getCachedMessage("msg2")).toEqual(msg2) // Still not considered
+  expect(getCachedMessage("msg3")).toEqual(msg3) // Still not considered
+
+  // Now run with no fragment IDs -> all should expire since
+  // we only allow an age of 1 after a fragment run (might be optimized at some point)
```

Comment:
yeah I think this is something we could probably optimize; 1 sounds to be pretty low

Reply:
Yep, I think we should tackle that as part of [this follow-up](https://www.notion.so/snowflake-corp/Investigate-a-better-forward-message-cache-cleanup-mechanism-1bd7170bb41680728e41c28db01baadc). Maybe the actual solution here is to just use an LRU cached with fixed total cache size so that we don't even have to count the message age.